### ISSUE TITLE
Ability to set custom stoploss before entry

### ISSUE THREAD

**Issue by Axel-CH** [freqtrade/freqtrade#9233]
## Context
Using **custom entry** or exit price, user can wait postpone his entry or exit when the market reach a price level.
It is a great tool to enter at key levels or market structure (last period low, last period high, Fibonacci retracement / extension, etc...)
However, for an entry case, if the market reach the desired price, but violently move deeper, there is a big loss potential, as the stop loss is not yet placed on exchange.

## Describe the enhancement
Being able to take advantage of the possibility to create entry stop loss directly within entry order could mitigate greatly that risk.

Why?
1) It will allow an increased risk protection on the position in case of sudden price change, as there is no gap in time between entry and stop-loss on exchange activation.

2) It will allow more precise stop loss based on user defined market structure.

*Explain the enhancement you would like*
Ability to set custom stop-loss in entry order

**Note**
This feature will be particularly useful if user want to enter at custom price, and exit immediately if a price level is not held.
Tested with ccxt and bybit with this sample code:

```
# SL set at order creation existing order
symbol = 'LOOM/USDT:USDT'
type = 'limit'  # or 'market'
side = 'buy'
amount = 11  # your amount
price = 0.10  # your price
params = {
    'stopLoss': {
        'type': 'limit',  # or 'market', this field is not necessary if limit price is specified
        'price': 0.09,  # limit price for a limit stop loss order
        'triggerPrice': 0.095,
    }
}
order = exchange.create_order (symbol, type, side, amount, price, params)
```

On exchange Order:
![entry_order_with_sl_lq](https://github.com/freqtrade/freqtrade/assets/7593404/251bcedd-40f9-47d3-8fa1-04325a57f2ce)

**Implementation idea:**
As `custom_stoploss` require an active Trade object, I would add a new callback named `initial_custom_stoploss` or `entry_custom_stoploss`

**Comment by xmatthias** [comment]
I don't think that this type of order placement has wide-spread support (so far, i've only seen it on bybit futures).

I only see this as worth the time if it's supported by multiple exchanges, in which case it can provide benefits to a wider userbase - so please let me know which of binance/okx/gate do support this (obviously using the same ccxt interface - and on which markets this support exists (spot, futures)).

Also, please investigate how this needs to be handled after the fact (like - how do we get the order-id for these additional orders - across the supported exchanges)?

i've briefly looked at this for bybit, and it seems that while the placement of the initial orders is pretty simple - you only get one order id back, so you essentially can't track what happens to the "other" orders once the initial order fills.

---

From an interface perspective - that's the wront point in time for this. 
first investigate the above points - which will lead to a decision about the "if" - then we can have a look the "how".

I've marked this as question for now - as i'm pretty sure it'll be a bybit only feature - but proof me wrong, and it may be a nice enhancement.

**Comment by Axel-CH** [comment]
You are right on the fact that this type of order is not a standard and It could be complicated to track the order once created within an entry.
I think we can achieve the same goal of having a stop-loss order on exchange before entry via different way:
Using a distinct stop market order or limit.
A majority of exchanges accept such order, as it is not a "real" order in order-book, but an instruction to add it if possible and at certain conditions. 

You can see on majority of exchange, for perpetual contracts, I can put theses trigger order before having an actual position, even an open order. This way we can have a separate "orderid" that will allow us to manage the order normally.
We just have to make sure to be in "reduceOnly" mode at each time. 

Exemples:
**OKX**
![trigger_exit_order_okx](https://github.com/freqtrade/freqtrade/assets/7593404/63aa322a-96cc-44cb-9dfb-bb897adcca07)
**Kucoin**
![trigger_exit_order_kucoin](https://github.com/freqtrade/freqtrade/assets/7593404/92b19943-ba7b-46d7-9a90-5a75f0afb16e)
**Bybit**
![trigger_exit_order_bybit](https://github.com/freqtrade/freqtrade/assets/7593404/510230b6-ebe5-4b2c-b801-b18a15069920)

**Comment by xmatthias** [comment]
it's a possibility eventually in the future - but you need to be VERY cautious to ensure orders are properly canceled alongside other orders / deleting the trade - which is all logic we currently don't need/have.
It'll also need logic / handling for markets / exchanges that don't support this (e.g. binance spot, where a stop order does consume the amount, and fails to place if you don't own it or it's tied up in another order).

i don't see the need for anything new though from an interface perspective.

The way i see the process:
* place the entry order
* trade object is created
* initial "stop" call is made
* stop order is created

that's a gap of milliseconds (depending on the exchange) - if you place your order to "not fill immediately" - it shouldn't be a problem
if it does - it's also not a realistic problem, as then you'd have to be "filled immediately, and market crashes within 1 second".

It'll however not happen soon ... things like stoploss_order_id will make this cumbersome and difficult to handle.

**Comment by snussik** [comment]
@Axel-CH @xmatthias  I'm using the following code snippet to set custom stoploss just after initial stoploss created. And If side and  stoploss value calculated right Initial stoploss doesn't affect much. 

<details>
<summary>Setting customs stoploss on entry code snippet</summary>

```python 

from dataclasses import dataclass
import datetime
from pandas import DataFrame

from freqtrade.persistence.trade_model import Order, Trade
from freqtrade.strategy.interface import IStrategy
import numpy as np

from enum import Enum


# ================================
# Helpers
# ================================
@dataclass
class StopLoss(Enum):
    SL = "sl"

def calc_sl_percent(current_rate: float, sl_value: float, trade_is_short: bool) -> float:
        if trade_is_short:
            return ((sl_value / current_rate) - 1)
        if not trade_is_short:
            return (1 - (sl_value / current_rate))


# ================================
# End Helpers
# ================================


def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

    some_condition = np.zeros(len(dataframe), dtype=bool)   

    # Setting stoploss value here
    dataframe.loc[some_condition, f"{StopLoss.SL.value}"] = dataframe.loc[some_condition, 'high'] # just for example

    return dataframe

def order_filled(self: IStrategy, pair: str, trade: Trade, order: Order, current_time: datetime, **kwargs) -> None:

        stoploss_key = StopLoss.SL.value

        # ================================

        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        if len(dataframe) == 0:
            return None
        last_candle = dataframe.iloc[-1].squeeze()
        sl = last_candle[f'{stoploss_key}']


        if (trade.nr_of_successful_entries == 1) and (order.ft_order_side == trade.entry_side):

            # Setting trade stoploss value here
            trade.set_custom_data(key=stoploss_key, value=sl)
  
        return None

def custom_stoploss(self: IStrategy, pair: str, trade: Trade, current_time: datetime,
                        current_rate: float, current_profit: float, after_fill: bool, **kwargs) -> float:

        stoploss_key = StopLoss.SL.value
        # ================================

        # Returning custom stoploss pct from set stoploss value to current value
        sl_percent = trade.leverage * calc_sl_percent(current_rate, trade.get_custom_data(key=stoploss_key), trade.is_short)
        return sl_percent


```

</details>

For my case the most strong impact to entry and further losses make fee rates as if stoploss is tight they can be twice greater.

**Comment by xmatthias** [comment]
oh absolutely - but we can't impact fees, really 😆 
without fees, many more strategies would become suddenly profitable 😆 

From an approach perspective - there's nothing wrong to your sample - but it's not what this issue is about directly ;)

**Comment by snussik** [comment]
@xmatthias 

> but it's not what this issue is about directly ;)

In some point I think that it's quite near this issue. As initial trade is created with global stoploss value, but just after creation (on the second iteration) it changes with custom stoploss value. And if it's set correctly (with respect to current_price and  custom_stoploss setting logic), it's practically the same as the custom initial stoploss.

**Comment by xmatthias** [comment]
well except - the issue is about placing a stoploss order before even placing the entry.
